interleaving/probabilistic.py

```diff
@@ -178,31 +172,53 @@ def _compute_scores(cls, ranking, clicks, tau=3.0, n=10**4):
         if 2 < len(ranking.lists):
             # [Schuth+, SIGIR 2015]
             R = [cls.Softmax(tau, R_j) for R_j in ranking.lists]
-            A_prime = [([0] * len(R), 0.0, [])]
+            A_prime = [(np.zeros(len(R)), 0.0, [])]
             threshold = 1 / len(R) * n ** (1 / len(L))
             for d in L:
-                A, A_prime = A_prime, []
-                P = [None] * len(R)
+                # Break if no click
+                # Stop when all the clicks are examined.
+                if len(C) == 0:
+                    break
+                d_in_C = d in C
+                if d_in_C:
+                    C.remove(d)
+
+                # Compute the document probability
+                # Only keep non-zero rankers
+                P = np.zeros(len(R))
+                R_non_zero = []
                 for j, R_j in enumerate(R):
                     P[j] = R_j.delete(d)
-                for (o, p, a) in A:
-                    for j, R_j in enumerate(R):
-                        if threshold < np.random.rand():
-                            continue
-                        if P[j] <= 0.0:
-                            continue
-                        p_prime = p + np.log(P[j] / 2)
-                        o_prime = o[:]
-                        if d in C:
+                    if P[j] > 0.0:
+                        R_non_zero.append((j, R_j))
+                if len(R_non_zero) == 0:
+                    continue
+
+                A, A_prime = A_prime, []
+                is_pass = np.random.rand(len(A), len(R_non_zero)) <= threshold
+                for i, (o, p, a) in enumerate(A):
+                    # Skip some assignments with certain probability
+                    R_used = [R_non_zero[k] for k in range(len(R_non_zero))
+                        if is_pass[i][k]]
+                    for j, R_j in R_used:
+                        p_prime = p + np.log(P[j])
```

Comment:
I have no idea why `/ 2` is removed.

Reply:
I assumed that the original version is already wrong.
I think 1/2 comes from Probabilistic Interleaving, where P(a|l, q) = P(l|a, q)P(a|q)/P(l|q) and P(a|q) = 1/(the number of rankers). P(a|q) = 1/2 for PI. Thus, 1/2 may represent P(a|q).

Although this part must be updated for PM, i.e. `1/len(R)`,
this term does not affect the final outcome and can be removed as follows:

`P(a|l, q) = P(l|a, q)P(a|q)/P(l|q) = P(l|a, q)P(a|q)/sum_a P(l|a, q)P(a|q) = P(l|a, q)/sum_a P(l|a, q)`

where P(a|q) = 1/(the number of rankers) and does not depend on `a`.

Reply:
I understand. Thank you!